File: frontend/playground/battle/calculator/target-calculator.ts
import type {Position} from "~/playground/data/common";
import type {BattleCalculations} from "~/playground/data/calculations";

export const TargetCalculator = {
  calculatePositions: function (vehicleId: number | undefined, battle: BattleCalculations): Position[] {
    let positions = Object.values(battle.model.vehicles)
        .filter(vehicle => !vehicleId || battle.allowedTarget(vehicle.id, vehicleId))
        .map(vehicle => vehicle.state.position)
    if (positions.length > 0) {
      return positions
    }

    positions = Object.values(battle.model.drones)
        .filter(drone => !drone.vehicleId || !vehicleId || battle.allowedTarget(drone.vehicleId, vehicleId))
        .map(drone => drone.state.position)
    if (positions.length > 0) {
      return positions
    }

    return Object.values(battle.model.missiles)
        .filter(missile => !vehicleId || battle.allowedTarget(missile.vehicleId, vehicleId))
        .map(missile => missile.state.position)
  }
}

File: backend/src/main/java/com/github/aadvorak/artilleryonline/battle/calculator/TargetCalculator.java
package com.github.aadvorak.artilleryonline.battle.calculator;

import com.github.aadvorak.artilleryonline.battle.calculations.BattleCalculations;
import com.github.aadvorak.artilleryonline.battle.common.Position;

import java.util.Set;
import java.util.stream.Collectors;

public class TargetCalculator {

    public static Set<Position> calculatePositions(Integer vehicleId, BattleCalculations battle) {
        var battleModel = battle.getModel();
        var positions = battleModel.getVehicles().values().stream()
                .filter(vehicleModel -> vehicleId == null || battle.allowedTarget(vehicleModel.getId(), vehicleId))
                .map(vehicleModel -> vehicleModel.getState().getPosition().getCenter())
                .collect(Collectors.toSet());
        if (!positions.isEmpty()) {
            return positions;
        }
        positions = battleModel.getDrones().values().stream()
                .filter(droneModel -> droneModel.getVehicleId() == null || vehicleId == null
                        || battle.allowedTarget(droneModel.getVehicleId(), vehicleId))
                .map(droneModel -> droneModel.getState().getPosition().getCenter())
                .collect(Collectors.toSet());
        if (!positions.isEmpty()) {
            return positions;
        }
        return battleModel.getMissiles().values().stream()
                .filter(missileModel -> vehicleId == null || battle.allowedTarget(missileModel.getVehicleId(), vehicleId))
                .map(missileModel -> missileModel.getState().getPosition().getCenter())
                .collect(Collectors.toSet());
    }
}
